### PR TITLE
Authenticate with GCM without GCE metadata

### PR DESCRIPTION
**Pull request by georgebuckerfield**
This adds functionality to authenticate against GCM without requiring the GCE metadata service. It instead uses the`GOOGLE_APPLICATION_CREDENTIALS` environment variable to lookup a credentials file downloaded from GCP. If this variable is not set, the process falls back to using the GCE metadata. 

This change allows users running clusters in AWS/on-prem to store heapster metrics in GCM/Stackdriver.

### CODE REVIEW

**Comment by georgebuckerfield** [comment]
I signed it!

**Comment by piosz** [comment]
@igorpeshansky from Stackdriver team.

**Comment by georgebuckerfield** [comment]
Is there anything else I need to do to move this along? Or is it unlikely to be merged? We're using a build of my forked version without any issues for a little while now but it would be great to switch back to the upstream build. Thanks!

File: metrics/sinks/gcm/gcm.go
Comment (igorpeshansky) [inline metrics/sinks/gcm/gcm.go]: There is another option -- the credentials file actually has the project id embedded in it. If the user didn't supply one, why not try the one from the credentials file, and then the one from the metadata server, in that order?

File: metrics/sinks/gcm/gcm.go
Comment (igorpeshansky) [inline metrics/sinks/gcm/gcm.go]: This is actually already done by `google.DefaultClient` when the credentials file isn't present. Looks like a bunch of that code can be reused.

**Review by igorpeshansky — CHANGES_REQUESTED** [review]

File: metrics/sinks/gcm/gcm.go
Comment (igorpeshansky) [inline metrics/sinks/gcm/gcm.go]: Two comments here:
(a) You want the project id from the credentials to override the one from the environment,
(b) You probably still want a way to supply the project id explicitly, in case the credentials aren't owned by the target project but simply have access to it.

File: metrics/sinks/gcm/gcm.go
Comment (igorpeshansky) [inline metrics/sinks/gcm/gcm.go]: Note: older credentials files may not have an explicit project id field, but have it embedded in the email address. Not sure it it's worth handling those cases here.

**Review by igorpeshansky — CHANGES_REQUESTED** [review]

File: metrics/sinks/gcm/gcm.go
Comment (georgebuckerfield) [inline metrics/sinks/gcm/gcm.go]: @igorpeshansky thanks for the feedback! I'll re-order where it looks for the project ID and add another option to provide it explicitly via an environment variable. So we'd check in this order:
- environment variable 
- credentials file
- GCE metadata

Does that sound right?

Comment (igorpeshansky) [inline metrics/sinks/gcm/gcm.go]: Yes, this is the order that makes the most sense from the Stackdriver perspective.

File: metrics/sinks/gcm/gcm.go
Comment (igorpeshansky) [inline metrics/sinks/gcm/gcm.go]: [Optional] The Google OAuth2 libraries also look for this file in the default location `/etc/google/auth/application_default_credentials.json`. @crassirostris and @piosz should decide if it makes sense to match this behavior.

**Review by igorpeshansky — CHANGES_REQUESTED** [review]

File: metrics/sinks/gcm/gcm.go
Comment (georgebuckerfield) [inline metrics/sinks/gcm/gcm.go]: I think that if the credentials are supplied in PK12 format, then `google.DefaultClient(oauth2.NoContext, gcm.MonitoringScope)` will fail earlier in the process? At least from testing it locally, I get:
```
error getting credentials using GOOGLE_APPLICATION_CREDENTIALS environment variable: invalid character '\u0082' after top-level value
```
if I point `GOOGLE_APPLICATION_CREDENTIALS` at a downloaded p12 file. 
But we could still have it fall back on the metadata server if there was a problem unmarshalling the json file for any other reason. 

I'm also thinking it would make sense to move some of this out into some functions in `common/gce/gce.go` to keep everything a bit tidier (`GetProjectIdFromEnv()`, `GetProjectIdFromFile()`, etc). 

Thanks again for taking the time to review this.

File: metrics/sinks/gcm/gcm.go
Comment (igorpeshansky) [inline metrics/sinks/gcm/gcm.go]: [Optional] Adding a trailing "." would be consistent with the above comments, but not the ones below... I'll let you pick the style you want to follow.

**Review by igorpeshansky — CHANGES_REQUESTED** [review]

File: metrics/sinks/gcm/gcm.go
Comment (igorpeshansky) [inline metrics/sinks/gcm/gcm.go]: That's fair -- the P12 format is listed as deprecated on the key download page.

I'll let the Kubernetes team comment on the refactoring, but keep in mind that extracting the project id from credentials is actually not GCE-specific (though it might be GCP-specific).

**Review by igorpeshansky — APPROVED** [review]
You'll need @crassirostris and @piosz to actually approve.

**Comment by crassirostris** [comment]
@georgebuckerfield Sorry for the delay. I actually think that your comment about moving this code to `k8s.io/heapster/common/gce` looks very reasonable, then in the target code you would have clean step

```
projectId, err := gce_utils.GetProjectId()
if err != nil {
  return nil, fmt.Errorf(...)
}
```

**Comment by georgebuckerfield** [comment]
No problem, @crassirostris - I'll work on moving that code this weekend, I think I have most of it in a separate branch already. Let me know if there's anything else obvious that needs looking at.

**Comment by crassirostris** [comment]
@georgebuckerfield Make code less nested, but I think that'll be obvious with a separate function :)

File: metrics/sinks/gcm/gcm.go
Comment (crassirostris) [inline metrics/sinks/gcm/gcm.go]: Why do you need those now?

File: common/gce/gce.go
Comment (crassirostris) [inline common/gce/gce.go]: I think you can just remove those two lines

File: common/gce/gce.go
Comment (crassirostris) [inline common/gce/gce.go]: Errors should not be capitalized. Check the rest of the code for this

**Review by crassirostris — CHANGES_REQUESTED** [review]

File: metrics/sinks/gcm/gcm.go
Comment (georgebuckerfield) [inline metrics/sinks/gcm/gcm.go]: My bad, forgot to remove them.

File: common/gce/gce.go
Comment (crassirostris) [inline common/gce/gce.go]: I think such comments should not be V(0), but rather V(4)

File: common/gce/gce.go
Comment (crassirostris) [inline common/gce/gce.go]: Let's be consistent with the rest of the code and use %v for formatting errors

**Review by crassirostris — CHANGES_REQUESTED** [review]
I think the overall readability will benefit from splitting out the main function into 3 high-level ones:

```
if projectId, err := tryGetFromEnvVar(); err != nil {
  return projectId, nil
} else {
  glog.V(4).Infof("... %v", err)
}

if projectId, err := tryGetFromFile(); err != nil {
  return projectId, nil
} else {
  glog.V(4).Infof("... %v", err)
}

projectId, err := tryGetFromMeta()
if err != nil {
  return projectId, nil
}
return "", fmt.Errorf("... %v", err)
```

**Review by crassirostris — APPROVED** [review]
I generally prefer %s for strings and %v for errors, but leave it to you :)

LGTM, please squash commits

**Comment by georgebuckerfield** [comment]
No, you're right, consistency with the rest of the code is better. I'll squash the commits when I get home tonight. Thanks for all your help with this!

**Comment by crassirostris** [comment]
@georgebuckerfield Thank you for your contribution!

Whoops, thought it's already being tested
/ok-to-test

**Comment by DirectXMan12** [comment]
Tests pass, has LGTM from sink maintainer.  Merging

**Pull request merged**
